htlcswitch/quiescer.go

```diff
@@ -0,0 +1,233 @@
+			ErrStfuAlreadySent, q.cfg.chanID)
+	}
+
+	if !q.canSendStfu() {
+		return fn.Errf[lnwire.Stfu]("%w for channel %v",
+			ErrPendingLocalUpdates, q.cfg.chanID)
+	}
+
+	stfu := lnwire.Stfu{
+		ChanID:    q.cfg.chanID,
+		Initiator: q.localInit,
+	}
+
+	return fn.Ok(stfu)
+}
+
+// oweStfu returns true if we owe the other party an Stfu. We owe the remote an
+// Stfu when we have received but not yet sent an Stfu, or we are the initiator
+// but have not yet sent an Stfu.
+func (q *quiescer) oweStfu() bool {
+	return q.received && !q.sent
+}
+
+// needStfu returns true if the remote owes us an Stfu. They owe us an Stfu when
+// we have sent but not yet received an Stfu.
+func (q *quiescer) needStfu() bool {
+	return q.sent && !q.received
+}
+
+// isQuiescent returns true if the state machine has been driven all the way to
+// completion. If this returns true, processes that depend on channel quiescence
+// may proceed.
+func (q *quiescer) isQuiescent() bool {
+	return q.sent && q.received
+}
+
+// quiescenceInitiator determines which ChannelParty is the initiator of
+// quiescence for the purposes of downstream protocols. If the channel is not
+// currently quiescent, this method will return ErrNoDownstreamLeader.
+func (q *quiescer) quiescenceInitiator() fn.Result[lntypes.ChannelParty] {
+	switch {
+	case !q.isQuiescent():
+		return fn.Err[lntypes.ChannelParty](ErrNoQuiescenceInitiator)
+
+	case q.localInit && q.remoteInit:
+		// In the case of a tie, the channel initiator wins.
+		return fn.Ok(q.cfg.channelInitiator)
+
+	case !q.localInit && !q.remoteInit:
+		// We assume it is impossible for both to be false if the
+		// channel is quiescent.
+		panic("impossible: one party must have initiated quiescence")
+
+	case q.localInit:
+		return fn.Ok(lntypes.Local)
+
+	case q.remoteInit:
+		return fn.Ok(lntypes.Remote)
+	}
+
+	panic("impossible: non-exhaustive switch quiescer.downstreamLeader")
+}
+
+// canSendUpdates returns true if we haven't yet sent an Stfu which would mark
+// the end of our ability to send updates.
+func (q *quiescer) canSendUpdates() bool {
+	return !q.sent && !q.localInit
+}
+
+// canRecvUpdates returns true if we haven't yet received an Stfu which would
+// mark the end of the remote's ability to send updates.
+func (q *quiescer) canRecvUpdates() bool {
+	return !q.received
+}
+
+// canSendStfu returns true if we can send an Stfu.
+func (q *quiescer) canSendStfu() bool {
+	return q.cfg.numPendingUpdates(lntypes.Local, lntypes.Local) == 0 &&
+		q.cfg.numPendingUpdates(lntypes.Local, lntypes.Remote) == 0
+}
+
+// canRecvStfu returns true if we can receive an Stfu.
+func (q *quiescer) canRecvStfu() bool {
+	return q.cfg.numPendingUpdates(lntypes.Remote, lntypes.Local) == 0 &&
+		q.cfg.numPendingUpdates(lntypes.Remote, lntypes.Remote) == 0
+}
+
+// drive drives the quiescence machine forward. It returns an error if the state
+// machine is in an invalid state.
+func (q *quiescer) drive() error {
```

Comment:
Seems it just handles sending, and not also the receiving case. So it's just a way to make a sending step forward, but doesn't necessarily give a caller a single point to drive the state machine forward based on input events.

Reply:
`maybeTriggerSftu`?

`maybeSendStfu`?

Reply:
renamed to `sendOwedStfu`